File: src/deb.rs
use std::collections::BTreeMap;
use std::io::{Cursor, Read};
use std::path::Path;

use anyhow::{Context, bail};
use bzip2::read::BzDecoder;
use flate2::read::GzDecoder;
use md5::Md5;
use sha1::Sha1;
use sha2::{Digest, Sha256};
use tar::Archive;
use xz2::read::XzDecoder;
use zstd::stream::read::Decoder as ZstdDecoder;

use crate::apt::parse_control_paragraphs;

#[derive(Debug, Clone)]
pub struct DebPackage {
    pub filename: String,
    pub size: u64,
    pub md5: String,
    pub sha1: String,
    pub sha256: String,
    pub control: BTreeMap<String, String>,
    pub raw: Vec<u8>,
}

impl DebPackage {
    pub fn from_path(path: &Path) -> anyhow::Result<Self> {
        let raw =
            std::fs::read(path).with_context(|| format!("failed to read {}", path.display()))?;
        let filename = path
            .file_name()
            .and_then(|f| f.to_str())
            .context("package path has no UTF-8 filename")?
            .to_string();
        Self::parse(filename, raw)
    }

    pub fn parse(filename: String, raw: Vec<u8>) -> anyhow::Result<Self> {
        let size = raw.len() as u64;
        let md5 = hex::encode(Md5::digest(&raw));
        let sha1 = hex::encode(Sha1::digest(&raw));
        let sha256 = hex::encode(Sha256::digest(&raw));
        let control = extract_control(&raw)?;

        Ok(Self {
            filename,
            size,
            md5,
            sha1,
            sha256,
            control,
            raw,
        })
    }

    pub fn package(&self) -> anyhow::Result<&str> {
        self.control
            .get("Package")
            .map(String::as_str)
            .context("package does not specify Package")
    }

    pub fn version(&self) -> anyhow::Result<&str> {
        self.control
            .get("Version")
            .map(String::as_str)
            .context("package does not specify Version")
    }

    pub fn architecture(&self) -> anyhow::Result<&str> {
        self.control
            .get("Architecture")
            .map(String::as_str)
            .context("package does not specify Architecture")
    }

    pub fn generated_filename(&self) -> anyhow::Result<String> {
        Ok(format!(
            "{}_{}_{}.deb",
            self.package()?,
            self.version()?,
            self.architecture()?
        ))
    }

    pub fn package_entry(&self, pool_path: &str) -> String {
        let mut out = String::new();
        for (key, value) in &self.control {
            out.push_str(&format!("{key}: {value}\n"));
        }
        out.push_str(&format!("Filename: {pool_path}\n"));
        out.push_str(&format!("Size: {}\n", self.size));
        out.push_str(&format!("MD5sum: {}\n", self.md5));
        out.push_str(&format!("SHA1: {}\n", self.sha1));
        out.push_str(&format!("SHA256: {}\n", self.sha256));
        out
    }
}

fn extract_control(data: &[u8]) -> anyhow::Result<BTreeMap<String, String>> {
    let mut reader = ArReader::new(data)?;
    while let Some(entry) = reader.next_entry()? {
        if entry.name.starts_with("control.tar") {
            return extract_control_from_tar(&entry.data, &entry.name);
        }
    }
    bail!("control.tar not found in deb package")
}

fn extract_control_from_tar(data: &[u8], name: &str) -> anyhow::Result<BTreeMap<String, String>> {
    let tar_data = decompress_control_tar(data, name)?;
    let mut archive = Archive::new(Cursor::new(tar_data));

    for entry in archive.entries()? {
        let mut entry = entry?;
        let path = entry.path()?;
        if path == Path::new("./control") || path == Path::new("control") {
            let mut control = String::new();
            entry.read_to_string(&mut control)?;
            let paragraphs = parse_control_paragraphs(&control)?;
            return paragraphs
                .into_iter()
                .next()
                .context("control file did not contain a paragraph");
        }
    }

    bail!("control file not found in {name}")
}

fn decompress_control_tar(data: &[u8], name: &str) -> anyhow::Result<Vec<u8>> {
    if name.ends_with(".gz") || data.starts_with(&[0x1f, 0x8b]) {
        read_all(GzDecoder::new(Cursor::new(data)))
    } else if name.ends_with(".xz") || data.starts_with(&[0xfd, b'7', b'z', b'X', b'Z']) {
        read_all(XzDecoder::new(Cursor::new(data)))
    } else if name.ends_with(".zst") || data.starts_with(&[0x28, 0xb5, 0x2f, 0xfd]) {
        read_all(ZstdDecoder::new(Cursor::new(data))?)
    } else if name.ends_with(".bz2") || data.starts_with(b"BZh") {
        read_all(BzDecoder::new(Cursor::new(data)))
    } else {
        Ok(data.to_vec())
    }
}

fn read_all(mut reader: impl Read) -> anyhow::Result<Vec<u8>> {
    let mut out = Vec::new();
    reader.read_to_end(&mut out)?;
    Ok(out)
}

struct ArReader<'a> {
    data: &'a [u8],
    pos: usize,
}

struct ArEntry {
    name: String,
    data: Vec<u8>,
}

impl<'a> ArReader<'a> {
    fn new(data: &'a [u8]) -> anyhow::Result<Self> {
        if !data.starts_with(b"!<arch>\n") {
            bail!("not an ar archive");
        }
        Ok(Self { data, pos: 8 })
    }

    fn next_entry(&mut self) -> anyhow::Result<Option<ArEntry>> {
        if self.pos >= self.data.len() {
            return Ok(None);
        }
        if self.pos + 60 > self.data.len() {
            bail!("truncated ar header");
        }

        let header = &self.data[self.pos..self.pos + 60];
        self.pos += 60;

        if &header[58..60] != b"`\n" {
            bail!("invalid ar header ending");
        }

        let name = std::str::from_utf8(&header[0..16])?
            .trim()
            .trim_end_matches('/')
            .to_string();
        let size: usize = std::str::from_utf8(&header[48..58])?
            .trim()
            .parse()
            .context("invalid ar member size")?;

        if self.pos + size > self.data.len() {
            bail!("truncated ar member {name}");
        }

        let entry_data = self.data[self.pos..self.pos + size].to_vec();
        self.pos += size;
        if size % 2 == 1 && self.pos < self.data.len() {
            self.pos += 1;
        }

        Ok(Some(ArEntry {
            name,
            data: entry_data,
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use flate2::Compression;
    use flate2::write::GzEncoder;
    use std::io::Write;
    use tar::{Builder, Header};

    #[test]
    fn parses_gzipped_control_tar() {
        let deb = test_deb(
            b"Package: demo\nVersion: 1.0.0\nArchitecture: amd64\nDescription: Demo package\n",
        );
        let package = DebPackage::parse("demo.deb".to_string(), deb).unwrap();
        assert_eq!(package.package().unwrap(), "demo");
        assert_eq!(package.version().unwrap(), "1.0.0");
        assert_eq!(package.architecture().unwrap(), "amd64");
        assert_eq!(
            package.generated_filename().unwrap(),
            "demo_1.0.0_amd64.deb"
        );
    }

    fn test_deb(control: &[u8]) -> Vec<u8> {
        let control_tar = gzipped_control_tar(control);
        let empty_tar = gzipped_control_tar(b"Package: ignored\n");
        let mut ar = b"!<arch>\n".to_vec();
        append_ar_member(&mut ar, "debian-binary", b"2.0\n");
        append_ar_member(&mut ar, "control.tar.gz", &control_tar);
        append_ar_member(&mut ar, "data.tar.gz", &empty_tar);
        ar
    }

    fn gzipped_control_tar(control: &[u8]) -> Vec<u8> {
        let mut tar = Vec::new();
        {
            let mut builder = Builder::new(&mut tar);
            let mut header = Header::new_gnu();
            header.set_path("./control").unwrap();
            header.set_size(control.len() as u64);
            header.set_mode(0o644);
            header.set_cksum();
            builder.append(&header, control).unwrap();
            builder.finish().unwrap();
        }

        let mut gz = GzEncoder::new(Vec::new(), Compression::default());
        gz.write_all(&tar).unwrap();
        gz.finish().unwrap()
    }

    fn append_ar_member(ar: &mut Vec<u8>, name: &str, data: &[u8]) {
        let header = format!(
            "{:<16}{:<12}{:<6}{:<6}{:<8}{:<10}`\n",
            format!("{name}/"),
            0,
            0,
            0,
            0o100644,
            data.len()
        );
        assert_eq!(header.len(), 60);
        ar.extend_from_slice(header.as_bytes());
        ar.extend_from_slice(data);
        if data.len() % 2 == 1 {
            ar.push(b'\n');
        }
    }
}

File: internal/deb/deb.go
package deb

import (
	"archive/tar"
	"bytes"
	"compress/bzip2"
	"compress/gzip"
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/klauspost/compress/zstd"
	"github.com/ulikunitz/xz"
)

type DebPackage struct {
	Filename string
	Size     uint64
	MD5      string
	SHA1     string
	SHA256   string
	Control  map[string]string
	Raw      []byte
}

func FromPath(path string) (*DebPackage, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read %s: %w", path, err)
	}
	filename := filepath.Base(path)
	return Parse(filename, raw)
}

func Parse(filename string, raw []byte) (*DebPackage, error) {
	size := uint64(len(raw))

	md5Sum := md5.Sum(raw)
	md5Str := hex.EncodeToString(md5Sum[:])

	sha1Sum := sha1.Sum(raw)
	sha1Str := hex.EncodeToString(sha1Sum[:])

	sha256Sum := sha256.Sum256(raw)
	sha256Str := hex.EncodeToString(sha256Sum[:])

	control, err := extractControl(raw)
	if err != nil {
		return nil, err
	}

	return &DebPackage{
		Filename: filename,
		Size:     size,
		MD5:      md5Str,
		SHA1:     sha1Str,
		SHA256:   sha256Str,
		Control:  control,
		Raw:      raw,
	}, nil
}

// packageNameRe matches valid Debian package names: at least two characters,
// starting alphanumeric, then lowercase letters, digits, '+', '-', '.'.
var packageNameRe = regexp.MustCompile(`^[a-z0-9][a-z0-9+.-]+$`)

// versionRe matches the characters permitted in a Debian version string
// (epoch, upstream version, and revision). It deliberately excludes path
// separators and anything that could escape a pool path.
var versionRe = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9+.~:-]*$`)

// archRe matches Debian architecture names such as amd64, arm64, all, i386.
var archRe = regexp.MustCompile(`^[a-z0-9][a-z0-9-]*$`)

func (d *DebPackage) Package() (string, error) {
	val, ok := d.Control["Package"]
	if !ok {
		return "", fmt.Errorf("package does not specify Package")
	}
	if !packageNameRe.MatchString(val) {
		return "", fmt.Errorf("invalid Package name %q", val)
	}
	return val, nil
}

func (d *DebPackage) Version() (string, error) {
	val, ok := d.Control["Version"]
	if !ok {
		return "", fmt.Errorf("package does not specify Version")
	}
	if !versionRe.MatchString(val) {
		return "", fmt.Errorf("invalid Version %q", val)
	}
	return val, nil
}

func (d *DebPackage) Architecture() (string, error) {
	val, ok := d.Control["Architecture"]
	if !ok {
		return "", fmt.Errorf("package does not specify Architecture")
	}
	if !archRe.MatchString(val) {
		return "", fmt.Errorf("invalid Architecture %q", val)
	}
	return val, nil
}

func (d *DebPackage) GeneratedFilename() (string, error) {
	pkg, err := d.Package()
	if err != nil {
		return "", err
	}
	ver, err := d.Version()
	if err != nil {
		return "", err
	}
	arch, err := d.Architecture()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s_%s_%s.deb", pkg, ver, arch), nil
}

// reservedEntryFields are computed by the repository from the artifact itself.
// A control field in the .deb must never be able to override them, or a crafted
// package could point apt at an arbitrary file or forge a checksum.
var reservedEntryFields = map[string]bool{
	"Filename": true,
	"Size":     true,
	"MD5sum":   true,
	"SHA1":     true,
	"SHA256":   true,
}

func (d *DebPackage) PackageEntry(poolPath string) string {
	var keys []string
	for k := range d.Control {
		if reservedEntryFields[k] {
			continue
		}
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var out strings.Builder
	for _, k := range keys {
		fmt.Fprintf(&out, "%s: %s\n", k, foldFieldValue(d.Control[k]))
	}
	fmt.Fprintf(&out, "Filename: %s\n", poolPath)
	fmt.Fprintf(&out, "Size: %d\n", d.Size)
	fmt.Fprintf(&out, "MD5sum: %s\n", d.MD5)
	fmt.Fprintf(&out, "SHA1: %s\n", d.SHA1)
	fmt.Fprintf(&out, "SHA256: %s\n", d.SHA256)
	return out.String()
}

// foldFieldValue neutralizes control-field injection: every embedded newline in
// a field value must be followed by a space or tab (RFC822 continuation) so the
// value cannot forge a new field or an entire package stanza in the Packages
// index. A bare newline (or a paragraph break) from a crafted .deb is re-folded
// into a continuation line.
func foldFieldValue(value string) string {
	if !strings.ContainsAny(value, "\n\r") {
		return value
	}
	normalized := strings.ReplaceAll(value, "\r\n", "\n")
	normalized = strings.ReplaceAll(normalized, "\r", "\n")
	lines := strings.Split(normalized, "\n")
	var out strings.Builder
	for i, line := range lines {
		if i > 0 {
			out.WriteString("\n")
			// Ensure the continuation is a valid folded line and never blank
			// (a blank line would terminate the stanza).
			if !strings.HasPrefix(line, " ") && !strings.HasPrefix(line, "\t") {
				out.WriteString(" ")
			}
			if strings.TrimSpace(line) == "" {
				out.WriteString(".")
			}
		}
		out.WriteString(line)
	}
	return out.String()
}

func extractControl(data []byte) (map[string]string, error) {
	entries, err := parseAr(data)
	if err != nil {
		return nil, err
	}

	for _, entry := range entries {
		if strings.HasPrefix(entry.Name, "control.tar") {
			return extractControlFromTar(entry.Data, entry.Name)
		}
	}
	return nil, fmt.Errorf("control.tar not found in deb package")
}

func extractControlFromTar(data []byte, name string) (map[string]string, error) {
	tarData, err := decompressControlTar(data, name)
	if err != nil {
		return nil, err
	}

	tr := tar.NewReader(bytes.NewReader(tarData))
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}

		normalizedPath := filepath.Clean(hdr.Name)
		if normalizedPath == "control" || normalizedPath == "./control" || normalizedPath == ".control" {
			controlData, err := io.ReadAll(tr)
			if err != nil {
				return nil, err
			}
			paragraphs, err := parseControlParagraphs(string(controlData))
			if err != nil {
				return nil, err
			}
			if len(paragraphs) == 0 {
				return nil, fmt.Errorf("control file did not contain a paragraph")
			}
			return paragraphs[0], nil
		}
	}

	return nil, fmt.Errorf("control file not found in %s", name)
}

func decompressControlTar(data []byte, name string) ([]byte, error) {
	var r io.Reader = bytes.NewReader(data)
	var err error

	if strings.HasSuffix(name, ".gz") || bytes.HasPrefix(data, []byte{0x1f, 0x8b}) {
		r, err = gzip.NewReader(r)
		if err != nil {
			return nil, err
		}
	} else if strings.HasSuffix(name, ".xz") || bytes.HasPrefix(data, []byte{0xfd, '7', 'z', 'X', 'Z'}) {
		r, err = xz.NewReader(r)
		if err != nil {
			return nil, err
		}
	} else if strings.HasSuffix(name, ".zst") || bytes.HasPrefix(data, []byte{0x28, 0xb5, 0x2f, 0xfd}) {
		var decoder *zstd.Decoder
		decoder, err = zstd.NewReader(r)
		if err != nil {
			return nil, err
		}
		defer decoder.Close()
		r = decoder
	} else if strings.HasSuffix(name, ".bz2") || bytes.HasPrefix(data, []byte("BZh")) {
		r = bzip2.NewReader(r)
	}

	return io.ReadAll(r)
}

type arEntry struct {
	Name string
	Data []byte
}

func parseAr(data []byte) ([]arEntry, error) {
	if len(data) < 8 || string(data[:8]) != "!<arch>\n" {
		return nil, fmt.Errorf("not an ar archive")
	}

	var entries []arEntry
	pos := 8
	for pos < len(data) {
		if pos+60 > len(data) {
			return nil, fmt.Errorf("truncated ar header")
		}
		header := data[pos : pos+60]
		pos += 60

		if string(header[58:60]) != "`\n" {
			return nil, fmt.Errorf("invalid ar header ending")
		}

		name := strings.TrimSpace(string(header[0:16]))
		name = strings.TrimSuffix(name, "/")

		sizeStr := strings.TrimSpace(string(header[48:58]))
		var size int
		_, err := fmt.Sscanf(sizeStr, "%d", &size)
		if err != nil {
			return nil, fmt.Errorf("invalid ar member size: %w", err)
		}

		if pos+size > len(data) {
			return nil, fmt.Errorf("truncated ar member %s", name)
		}

		entryData := data[pos : pos+size]
		pos += size
		if size%2 == 1 && pos < len(data) {
			pos++
		}

		entries = append(entries, arEntry{
			Name: name,
			Data: entryData,
		})
	}
	return entries, nil
}

func parseControlParagraphs(content string) ([]map[string]string, error) {
	if strings.TrimSpace(content) == "" {
		return nil, nil
	}

	normalized := strings.ReplaceAll(content, "\r\n", "\n")
	paragraphs := strings.Split(normalized, "\n\n")

	var result []map[string]string
	for _, p := range paragraphs {
		if strings.TrimSpace(p) == "" {
			continue
		}
		fields, err := parseParagraph(p)
		if err != nil {
			return nil, err
		}
		result = append(result, fields)
	}
	return result, nil
}

func parseParagraph(paragraph string) (map[string]string, error) {
	fields := make(map[string]string)
	var currentKey string

	lines := strings.Split(paragraph, "\n")
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			// Skip empty lines in the middle
			continue
		}

		if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
			if currentKey == "" {
				return nil, fmt.Errorf("control continuation line without a preceding field")
			}
			fields[currentKey] = fields[currentKey] + "\n" + line
			continue
		}

		parts := strings.SplitN(line, ":", 2)
		if len(parts) != 2 {
			return nil, fmt.Errorf("invalid control field line: %s", line)
		}

		key := strings.TrimSpace(parts[0])
		if key == "" {
			return nil, fmt.Errorf("empty control field name")
		}
		value := strings.TrimLeft(parts[1], " \t")
		fields[key] = value
		currentKey = key
	}

	return fields, nil
}
